java-security/src/main/java/com/sap/cloud/security/token/validation/validators/OAuth2TokenKeyServiceWithCache.java

```diff
@@ -161,32 +138,43 @@ public PublicKey getPublicKey(JwtSignatureAlgorithm keyAlgorithm, String keyId,
 	 *             in case the algorithm of the json web key is not supported.
 	 *
 	 */
-	public PublicKey getPublicKey(JwtSignatureAlgorithm keyAlgorithm, String keyId, URI keyUri, String appTid, String clientId)
+	public PublicKey getPublicKey(JwtSignatureAlgorithm keyAlgorithm, String keyId, URI keyUri, String appTid, String clientId, String azp)
 			throws OAuth2ServiceException, InvalidKeySpecException, NoSuchAlgorithmException {
 		assertNotNull(keyAlgorithm, "keyAlgorithm must not be null.");
 		assertHasText(keyId, "keyId must not be null.");
 		assertNotNull(keyUri, "keyUrl must not be null.");
 
-		JsonWebKeySet keySet = getCache().getIfPresent(keyUri.toString());
-		if (keySet == null || !keySet.containsAppTid(appTid)) {
-			keySet = retrieveTokenKeysAndUpdateCache(keyUri, appTid, keySet, clientId); // creates and updates cache entries
-		}
-		if (keySet == null || keySet.getAll().isEmpty()) {
-			LOGGER.error("Retrieved no token keys from {}", keyUri);
+		CacheKey cacheKey = new CacheKey(keyUri, appTid, clientId, azp);
```

Comment:
are you sure all these parameters have to be included in the cache key? would JWKs be different for each possible parameter combination?